torch/distributed/fsdp/sharded_grad_scaler.py

```diff
@@ -0,0 +1,329 @@
+            # For the FSDP + Mixed Precision use case, the loss output is in the Mixed Precision
+            # format (fp16, bf16) and so the scaled loss should be of the same dtype.
+            return scaled_output.type(outputs.dtype)
+
+        stash: List[_GeneralMultiDeviceReplicator] = []
+
+        def apply_scale(val: Union[torch.Tensor, abc.Iterable[torch.Tensor]]) -> Union[torch.Tensor, abc.Iterable[torch.Tensor]]:
+            if isinstance(val, torch.Tensor):
+                assert _is_supported_device(val)
+                if len(stash) == 0:
+                    if self._scale is None:
+                        self._lazy_init_scale_growth_tracker(val.device)
+                    assert self._scale is not None
+                stash.append(_GeneralMultiDeviceReplicator(self._scale))
+                scaled_val = val * stash[0].get(val.device)
+                # Here we ensure the return dtype is the same as the outputs dtype.
+                # For the FSDP + Mixed Precision use case, the loss output is in the Mixed Precision
+                # format (fp16, bf16) and so the scaled loss should be of the same dtype.
+                return scaled_val.type(val.dtype)
+            elif isinstance(val, abc.Iterable):
+                iterable = map(apply_scale, val)
+                if isinstance(val, (list, tuple)):
+                    return type(val)(iterable)
+                else:
+                    return iterable
+            else:
+                raise ValueError("outputs must be a Tensor or an iterable of Tensors")
+
+        return apply_scale(outputs)
+
+    def _foreach_non_finite_check_and_unscale_cpu_(
+        self, grads: List, found_inf: torch.Tensor, inv_scale: torch.Tensor
+    ) -> None:
+        if len(grads) == 0:
+            return
+        assert inv_scale.numel() == 1, "inv_scale must be a 1-element tensor."
+        assert found_inf.numel() == 1, "found_inf must be a 1-element tensor."
+
+        expected_device = grads[0].device
+        for grad in grads:
+            for tensor in grad:
+                try:
+                    assert tensor.device == expected_device, "grads must be on the same device"
+                except AssertionError:
+                    logging.error("tensor device is %s and expected device is %s" % (tensor.device, expected_device))
+
+                # check for non_overlapping_and_dense doesn't exist in the python world
+                # as remarked here https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/AmpKernels.cu#L108
+                # we assume tensor is not MTA(multi tensor apply) safe. iterate through each item regardless of dtype
+                # TODO faster logic?
+                if torch.isinf(tensor).any().item() is True or torch.isnan(tensor).any().item() is True:  # type: ignore
+                    found_inf.data = torch.tensor([1.0])
+                    break
```

Comment:
this means some grads might be scaled and some might not be? I guess that is fine, because when we find inf, opt.step() will be a noop, and eventually we just won't use these grads basically right?

Reply:
Yup correct. Once we find inf/nan, opt.step is skipped.